algorithms.py

```diff
@@ -521,6 +684,113 @@ def _make_segment(self, left, right, count):
         return seg
 
 
+class Hull:
```

Comment:
A few comments would help here on these classes. Briefly, what are they and what are they for?